File: library/src/main/java/com/snail/commons/utils/Logger.java
package com.snail.commons.utils;

import android.os.Build;
import android.support.annotation.NonNull;
import android.util.Log;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.reflect.Field;

public class Logger {
    public static final int NONE = 1;
    public static final int VERBOSE = NONE << 1;
    public static final int DEBUG = VERBOSE << 1;
    public static final int INFO = DEBUG << 1;
    public static final int WARN = INFO << 1;
    public static final int ERROR = WARN << 1;
    public static final int ALL = VERBOSE | INFO | DEBUG | WARN | ERROR;

    public interface Filter {
        boolean accept(@NonNull String log);
    }

    private static boolean isSaveEnabled = false;
    private static int printLevel = NONE;
    private static Filter filter;

    /**
     * 控制输出级别<br>{@link #NONE}, {@link #VERBOSE}, {@link #DEBUG}, {@link #INFO}, {@link #WARN}, {@link #ERROR}
     */
    public static void setPrintLevel(int printLevel) {
        Logger.printLevel = printLevel;
    }

    public static void setFilter(Filter filter) {
        Logger.filter = filter;
    }

    /**
     * 控制是否执行saveLog方法
     */
    public static void setSaveEnabled(boolean isSaveEnabled) {
        Logger.isSaveEnabled = isSaveEnabled;
    }

    private static boolean accept(int priority, @NonNull String msg) {
        int level = getLevel(priority);
        return (printLevel & NONE) != NONE && (printLevel & level) == level && (filter == null || filter.accept(msg));
    }

    private static int getLevel(int priority) {
        switch (priority) {
            case Log.ERROR:
                return ERROR;
            case Log.WARN:
                return WARN;
            case Log.INFO:
                return INFO;
            case Log.DEBUG:
                return DEBUG;
            case Log.VERBOSE:
                return VERBOSE;
            default:
                return NONE;
        }
    }

    public static void v(String tag, String msg) {
        if (accept(Log.VERBOSE, msg)) {
            Log.v(tag, msg);
        }
    }

    public static void v(String tag, String msg, Throwable t) {
        if (accept(Log.VERBOSE, msg)) {
            Log.v(tag, msg, t);
        }
    }

    public static void d(String tag, String msg) {
        if (accept(Log.DEBUG, msg)) {
            Log.d(tag, msg);
        }
    }

    public static void d(String tag, String msg, Throwable t) {
        if (accept(Log.DEBUG, msg)) {
            Log.d(tag, msg, t);
        }
    }

    public static void i(String tag, String msg) {
        if (accept(Log.INFO, msg)) {
            Log.i(tag, msg);
        }
    }

    public static void i(String tag, String msg, Throwable t) {
        if (accept(Log.INFO, msg)) {
            Log.i(tag, msg, t);
        }
    }

    public static void w(String tag, String msg) {
        if (accept(Log.WARN, msg)) {
            Log.w(tag, msg);
        }
    }

    public static void w(String tag, String msg, Throwable t) {
        if (accept(Log.WARN, msg)) {
            Log.w(tag, msg, t);
        }
    }

    public static void e(String tag, String msg) {
        if (accept(Log.ERROR, msg)) {
            Log.e(tag, msg);
        }
    }

    public static void e(String tag, String msg, Throwable t) {
        if (accept(Log.ERROR, msg)) {
            Log.e(tag, msg, t);
        }
    }

    public static void saveLog(File file, String log) {
        if (isSaveEnabled) {
            try {
                BufferedWriter out = new BufferedWriter(new FileWriter(file, true));
                out.write(log);
                out.newLine();
                out.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public static void saveLog(File file, String appVersion, Throwable t) {
        if (isSaveEnabled) {
            try {
                StringWriter sw = new StringWriter();
                PrintWriter pw = new PrintWriter(sw);
                //获取手机的环境
                Field[] fields = Build.class.getDeclaredFields();
                pw.println("ERROR_OCCURRENCE_TIME=" + DateUtils.formatDate(System.currentTimeMillis(), "yyyy-MM-dd HH:mm:ss.SSS"));
                for (Field field : fields) {
                    field.setAccessible(true);
                    pw.println(field.getName() + "=" + field.get(null));
                }
                pw.println("AppVersion=" + appVersion);
                t.printStackTrace(pw);
                pw.println("\n");
                FileOutputStream fos = new FileOutputStream(file, true);
                fos.write(sw.toString().getBytes());
                fos.close();
                pw.close();
                sw.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: library/src/main/java/com/snail/commons/utils/Logger.kt
package com.snail.commons.utils

import android.os.Build
import android.util.Log
import java.io.*

object Logger {
    const val NONE = 1
    const val VERBOSE = NONE shl 1
    const val DEBUG = VERBOSE shl 1
    const val INFO = DEBUG shl 1
    const val WARN = INFO shl 1
    const val ERROR = WARN shl 1
    const val ALL = VERBOSE or INFO or DEBUG or WARN or ERROR

    private var isSaveEnabled = false
    private var printLevel = NONE
    private var filter: Filter? = null

    interface Filter {
        fun accept(tag: String, log: String): Boolean
    }

    /**
     * 控制输出级别.[NONE], [VERBOSE], [DEBUG], [INFO], [WARN], [ERROR]
     */
    fun setPrintLevel(printLevel: Int) {
        Logger.printLevel = printLevel
    }

    fun setFilter(filter: Filter) {
        Logger.filter = filter
    }

    /**
     * 控制是否执行saveLog方法
     */
    fun setSaveEnabled(isSaveEnabled: Boolean) {
        Logger.isSaveEnabled = isSaveEnabled
    }

    private fun accept(priority: Int, tag: String, msg: String): Boolean {
        val level = getLevel(priority)
        return printLevel and NONE != NONE && printLevel and level == level && (filter == null || filter!!.accept(tag, msg))
    }

    private fun getLevel(priority: Int): Int {
        return when (priority) {
            Log.ERROR -> ERROR
            Log.WARN -> WARN
            Log.INFO -> INFO
            Log.DEBUG -> DEBUG
            Log.VERBOSE -> VERBOSE
            else -> NONE
        }
    }

    fun v(tag: String, msg: String) {
        if (accept(Log.VERBOSE, tag, msg)) {
            Log.v(tag, msg)
        }
    }

    fun v(tag: String, msg: String, t: Throwable) {
        if (accept(Log.VERBOSE, tag, msg)) {
            Log.v(tag, msg, t)
        }
    }

    fun d(tag: String, msg: String) {
        if (accept(Log.DEBUG, tag, msg)) {
            Log.d(tag, msg)
        }
    }

    fun d(tag: String, msg: String, t: Throwable) {
        if (accept(Log.DEBUG, tag, msg)) {
            Log.d(tag, msg, t)
        }
    }

    fun i(tag: String, msg: String) {
        if (accept(Log.INFO, tag, msg)) {
            Log.i(tag, msg)
        }
    }

    fun i(tag: String, msg: String, t: Throwable) {
        if (accept(Log.INFO, tag, msg)) {
            Log.i(tag, msg, t)
        }
    }

    fun w(tag: String, msg: String) {
        if (accept(Log.WARN, tag, msg)) {
            Log.w(tag, msg)
        }
    }

    fun w(tag: String, msg: String, t: Throwable) {
        if (accept(Log.WARN, tag, msg)) {
            Log.w(tag, msg, t)
        }
    }

    fun e(tag: String, msg: String) {
        if (accept(Log.ERROR, tag, msg)) {
            Log.e(tag, msg)
        }
    }

    fun e(tag: String, msg: String, t: Throwable) {
        if (accept(Log.ERROR, tag, msg)) {
            Log.e(tag, msg, t)
        }
    }

    fun saveLog(file: File, log: String) {
        if (isSaveEnabled) {
            try {
                val out = BufferedWriter(FileWriter(file, true))
                out.write(log)
                out.newLine()
                out.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
    }

    fun saveLog(file: File, appVersion: String, t: Throwable) {
        if (isSaveEnabled) {
            try {
                val sw = StringWriter()
                val pw = PrintWriter(sw)
                //获取手机的环境
                val fields = Build::class.java.declaredFields
                pw.println("ERROR_OCCURRENCE_TIME=" + DateUtils.formatDate(System.currentTimeMillis(), "yyyy-MM-dd HH:mm:ss.SSS"))
                for (field in fields) {
                    field.isAccessible = true
                    pw.println(field.name + "=" + field.get(null))
                }
                pw.println("AppVersion=$appVersion")
                t.printStackTrace(pw)
                pw.println("\n")
                val fos = FileOutputStream(file, true)
                fos.write(sw.toString().toByteArray())
                fos.close()
                pw.close()
                sw.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}
